src/logics/ArithLogic.cc

```diff
@@ -0,0 +1,1304 @@
+}
+
+bool ArithLogic::isIntegralLogic() const {
+    switch (arithType) {
+        case ArithType::LIA:
+        case ArithType::NIA:
+        case ArithType::IDL:
+            return true;
+        default:
+            return false;
+    }
+}
+
+bool ArithLogic::isRealLogic() const {
+    switch (arithType) {
+        case ArithType::LRA:
+        case ArithType::NRA:
+            return true;
+        default:
+            return false;
+    }
+}
+
+void ArithLogic::checkArithSortCompatibleAndDefined(PTRef tr) const {
+    if (isIntegralLogic() and getSortRef(tr) != sort_INT) {
+        throw OsmtApiException("Incompatible sort");
+    } else if (isRealLogic() and getSortRef(tr) != sort_REAL) {
+        throw OsmtApiException("Incompatible sort");
+    } else if (not isIntegralLogic() and not isRealLogic()) {
+        throw OsmtApiException("Sort not uniquely defined");
+    }
+}
+
+SymRef ArithLogic::get_sym_Num_PLUS() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_PLUS : sym_Real_PLUS;
+}
+
+SymRef ArithLogic::get_sym_Num_MINUS() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_MINUS : sym_Real_MINUS;
+}
+
+const SymRef ArithLogic::get_sym_Num_NEG() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_NEG : sym_Real_NEG;
+}
+
+const SymRef ArithLogic::get_sym_Num_LEQ() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_LEQ : sym_Real_LEQ;
+}
+
+const SymRef ArithLogic::get_sym_Num_GEQ() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_GEQ : sym_Real_GEQ;
+}
+
+const SymRef ArithLogic::get_sym_Num_LT() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_LT : sym_Real_LT;
+}
+
+const SymRef ArithLogic::get_sym_Num_GT() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_GT : sym_Real_GT;
+}
+
+const SymRef ArithLogic::get_sym_Num_EQ() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_EQ: sym_Real_EQ;
+}
+
+const SymRef ArithLogic::get_sym_Num_ZERO() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_ZERO : sym_Real_ZERO;
+}
+
+const SymRef ArithLogic::get_sym_Num_ONE() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_ONE : sym_Real_ONE;
+}
+
+const SymRef ArithLogic::get_sym_Num_ITE() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_ITE : sym_Real_ITE;
+}
+
+const SymRef ArithLogic::get_sym_Num_ABS() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_ABS : sym_Real_ABS;
+}
+
+SymRef ArithLogic::get_sym_Num_TIMES() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_TIMES : sym_Real_TIMES;
+}
+
+SymRef ArithLogic::get_sym_Num_NTIMES() const {
+    checkArithSortDefined();
+    return isIntegralLogic() ? sym_Int_NTIMES : sym_Real_NTIMES;
+}
+
+bool ArithLogic::isNegated(PTRef tr) const {
```

Comment:
Maybe it could go to `LAVarMapper`?  If I move it to `LASolver` I can't write an assertion in `LAVarMapper::RegisterNewMapping` that the argument is not negated.  And also conceptually it's a thing that you could expect `LAVarMapper` to know.